src/twisted/internet/defer.py

```diff
@@ -198,16 +223,61 @@ def setDebugging(on):
     Deferred.debug = bool(on)
 
 
-def getDebugging():
+def getDebugging() -> bool:
     """
     Determine whether L{Deferred} debugging is enabled.
     """
     return Deferred.debug
 
 
+def _cancelledToTimedOutError(value: _T, timeout: float) -> _T:
+    """
+    A default translation function that translates L{Failure}s that are
+    L{CancelledError}s to L{TimeoutError}s.
+
+    @param value: Anything
+    @param timeout: The timeout
+
+    @raise: L{TimeoutError}
+
+    @since: 16.5
+    """
+    if isinstance(value, Failure):
+        value.trap(CancelledError)
+        raise TimeoutError(timeout, "Deferred")
+    return value
+
+
+# type note: this should be Callable[[object, ...], object] but mypy doesn't allow
+#   Callable[[object], object] is next best, but disallows valid callback signatures
+DeferredCallback = Callable[..., object]
+# type note: this should be Callable[[Failure, ...], object] but mypy doesn't allow
+#   Callable[[Failure], object] is next best, but disallows valid callback signatures
+DeferredErrback = Callable[..., object]
```

Comment:
These two are used in public type signatures and I thought they would be useful to clients, so I didn't prefix them with `_`… but I'm not super happy with these signatures, since they aren't providing a good descriptions due to the limitations of `Callable`.

I think they remain useful in that when we eventually fix these, the user of these variables will then benefit from the checking.  I used the "next best" examples above when doing the initial typing work do distinguish callbacks from errbacks, for example, but we can't ship with those definitions because they don't allows for callbacks with extra arguments.

Reply:
does this work better?

```python
class DeferredErrback(Protocol):
    def __call__(v: Failure, *args: object) -> object: ...
```

Reply:
Interesting… thanks, I'll try that

Reply:
OK here is what I tried for `DeferredErrback`:

```python
class DeferredErrback(Protocol):
    def __call__(self, failure: Failure, *args: object) ->object:
        ...
```

That yields some errors.  One class of those is from cases where a single function is being used as both a callback and errback (for example: `passthru`), and I think those just need to be cast.

Then we have these, and I'm wondering if this is due to the protocol being defined incorrectly, since I *think* these should match:

```console
src/twisted/internet/defer.py:545:46: error: Argument 2 to "addCallbacks" of "Deferred" has incompatible type "Callable[[Union[Failure, BaseException, None]], None]"; expected "DeferredErrback"  [arg-type]
            return self.addCallbacks(d.callback, d.errback)
                                                 ^
src/twisted/internet/defer.py:903:41: error: Argument 2 to "addCallbacks" of "Deferred" has incompatible type "Callable[[Failure], None]"; expected "DeferredErrback"  [arg-type]
            self.addCallbacks(maybeSucceed, maybeFail)
                                            ^
src/twisted/internet/defer.py:1652:23: error: Argument 1 to "addErrback" of "Deferred" has incompatible type "Callable[[Failure], Deferred]"; expected "DeferredErrback"  [arg-type]
            it.addErrback(handleCancel)
                          ^
```

Reply:
I haven't had time to try this out yet, but I believe what you want here is something like

`DeferredCallback = TypeVar("DeferredCallback", bound=DeferredErrbackCallableProtocol)`

This doesn't quite address the binding of the result-type typevar into the callback chain.

I checked Pilot's temporary Deferred stubs, and apparently we just `Any` away this problem, so I think we can start off pretty lax here and tighten up in subsequent PRs if it proves too much of a puzzle.